src/main/java/org/prebid/server/bidder/loyal/LoyalBidder.java

```diff
@@ -0,0 +1,155 @@
+import com.iab.openrtb.response.BidResponse;
+import com.iab.openrtb.response.SeatBid;
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.appush.proto.AppushImpExtBidder;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderCall;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
+import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
+import org.prebid.server.proto.openrtb.ext.request.loyal.ExtImpLoyal;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.BidderUtil;
+import org.prebid.server.util.HttpUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class LoyalBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpLoyal>> LOYAL_EXT_TYPE_REFERENCE =
+            new TypeReference<>() {
+            };
+
+    private static final String PUBLISHER_PROPERTY = "publisher";
+    private static final String NETWORK_PROPERTY = "network";
+    private static final String BIDDER_PROPERTY = "bidder";
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
+
+    public LoyalBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        final List<HttpRequest<BidRequest>> httpRequests = new ArrayList<>();
+
+        for (Imp imp : request.getImp()) {
+            final ExtImpLoyal extImpLoyal;
+            try {
+                extImpLoyal = parseExtImp(imp);
+            } catch (PreBidException e) {
+                return Result.withError(BidderError.badInput(e.getMessage()));
+            }
+
+            final Imp modifiedImp = modifyImp(imp, extImpLoyal);
+            httpRequests.add(makeHttpRequest(request, modifiedImp));
+        }
+
+        return Result.withValues(httpRequests);
+    }
+
+    private ExtImpLoyal parseExtImp(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), LOYAL_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private AppushImpExtBidder resolveImpExt(ExtImpLoyal extImpLoyal) {
+        final AppushImpExtBidder.AppushImpExtBidderBuilder builder = AppushImpExtBidder.builder();
```

Comment:
please use models related to the LoyalBidder only, each bidder is a separate and independent piece of logic

Reply:
this is only proto. I suggest MgidxBidder

Reply:
okay, please ping me when you the PR is ready for the review

Reply:
I suggested mgdixBidder. The class is written like that there. Is it wrong? The same is written for the error list. 
I think I got a little confused, first I was supposed to look at the betweenBidder, voxBidder then bematterfullBidder and now MgdixBidder. I'm confused

Reply:
The porting goal is implementing the same logic the the GO version has. The Loyal bidder might have similarities with existing bidders, you can use them as inspiration and/or examples, but the goal is replicating the logic of the Loyal Bidder from PBS Go (https://github.com/prebid/prebid-server/pull/3586).